### PR TITLE
Core: Avoid framerate dropping down 5x if frames can't be handled in time.

### PR DESCRIPTION
**Pull request by adrian17**
See comment in commit for details.

Currently Ruffle turns every movie that's consistently "slightly slower than expected FPS" into "5x worse than expected FPS".
z0r.de already fixed it on its fork by force-setting `MAX_FRAMES_PER_TICK = 1`. This patch basically does the same thing, but smarter - Ruffle still can run multiple "frames" at a time, but only if it thinks that based on past `run_frame()` timings, it can afford it within a single frame. This was @Herschel 's idea :)
In practice, the proposed video decoders are slow enough on wasm that `max_frames_per_tick()` will always return 1 for 99% of content. Same with most games (except, for example, Fancy Pants).

Oh, and I didn't test this on desktop; just on web.

### CODE REVIEW

File: core/src/player.rs [hunk 1]
```diff
@@ -188,6 +189,7 @@ pub struct Player {
     /// This is how we support custom SWF framerates
     /// and compensate for small lags by "catching up" (up to MAX_FRAMES_PER_TICK).
     frame_accumulator: f64,
+    recent_run_frame_timings: CircularQueue<f64>,
```
Comment (Herschel) [inline core/src/player.rs:192]: Why not `std::collections::VecDeque`?

File: core/src/player.rs [hunk 2]
```diff
@@ -436,6 +439,30 @@ impl Player {
...
+    /// This logic is far from perfect, as it doesn't take into account
+    /// that things like rendering also take time. But for now it's good enough.
+    fn max_frames_per_tick(&self) -> u32 {
+        if self.recent_run_frame_timings.is_empty() {
+            5
+        } else {
+            let frame_time = 1000.0 / self.frame_rate;
+            let average_run_frame_time = self.recent_run_frame_timings.iter().sum::<f64>()
+                / self.recent_run_frame_timings.len() as f64;
+            ((frame_time / average_run_frame_time) as u32).max(1).min(5)
```
Comment (Herschel) [inline core/src/player.rs:462]: Let's move `const MAX_FRAMES_PER_TICK: u32 = 5;` here and use it.

File: core/src/player.rs (hunk 1 above)
Comment (adrian17) [reply core/src/player.rs:192]: Then you need something like

```
self.recent_run_frame_timings.push_back(elapsed);
if self.recent_run_frame_timings.len() > 10 {
    self.recent_run_frame_timings.pop_front();
}
```
, right?
That's what I had originally, and I felt like using a data structure that "rotates on push" automatically would be a bit more elegant.
I can go back to `VecDeque` if you want ofc.

Comment (Herschel) [reply core/src/player.rs:192]: Yea, I'd rather avoid an extra dependency in this case. Thank you!

**Comment by Herschel** [comment]
@torokati44 Could you try this with your video content and see if it helps?

File: core/src/player.rs (hunk 1 above)
Comment (adrian17) [reply core/src/player.rs]: Done.

**Comment by torokati44** [comment]
Yep, it definitely does.

This is with current `master`:

https://user-images.githubusercontent.com/288816/106826892-1c38c500-6688-11eb-8a8d-80e0e4643fb7.mp4

And with this patch:

https://user-images.githubusercontent.com/288816/106826959-3e324780-6688-11eb-9573-9df7d68cc712.mp4

In this case, it does exactly what my original "fix" did, as `max_frames_per_tick` always returns `1`.
Still lags behind the sound (not included) considerably, but at least it's not as choppy meanwhile.

(Loop is: https://z0r.de/4449. I had to run it in debug mode tho, as on Desktop, decoding is _much_ faster anyway...)

**Comment by Herschel** [comment]
Thank you!

**Pull request merged**